src/completion.ts

```diff
@@ -0,0 +1,239 @@
+    }
+
+    const index = Number(numberString);
+    if (
+      !numberString || !Number.isFinite(index) || !Number.isInteger(index) ||
+      index < 0
+    ) {
+      throw new Error(`Expected an integer, found ${numberString}`);
+    }
+
+    if (!expect.includes(letter)) {
+      throw new Error(`Expected ${expect}, found ${letter}`);
+    }
+
+    if (letter === "s") {
+      const foundCommand = cmd.subcommands?.[index];
+      if (!foundCommand) {
+        throw new Error(
+          `No subcommand at index ${index} (${lookupPathString})`,
+        );
+      }
+      cmd = foundCommand;
+      continue;
+    }
+
+    if (letter === "o") {
+      const foundOption = cmd.options?.[index];
+      if (!foundOption) {
+        throw new Error(`No option at index ${index} (${lookupPathString})`);
+      }
+      option = foundOption;
+      expect = "a";
+      continue;
+    }
+
+    if (letter === "a") {
+      const foundArg = makeArray((option || cmd).args)?.[index];
+      console.log("foundArg", foundArg);
+      if (!foundArg) {
+        throw new Error(`No arg at index ${index} (${lookupPathString})`);
+      }
+      expect = "g";
+      arg = foundArg;
+      continue;
+    }
+
+    if (letter === "g") {
+      const foundGenerator = makeArray(arg?.generators)?.[index];
+      console.log("foundGenerator", makeArray);
+      if (!foundGenerator) {
+        throw new Error(`No generator at index ${index} (${lookupPathString})`);
+      }
+      generator = foundGenerator;
+      break;
+    }
+  }
+  if (!generator) {
+    throw new Error(
+      `Expected a generator, but found nothing (${lookupPathString})`,
+    );
+  }
+  return generator;
+}
+
+function kv(key: string, value: unknown): string {
+  return `${key}:${value}`;
+}
+
+function objectify(arr: string[]): string {
+  return `{${arr.join(",")}}`;
+}
+
+function arrayify(arr: string[]): string {
+  return `[${arr.join(",")}]`;
+}
+
+function kvObject(object: Record<string, unknown>): string {
+  const keyValues = Object.entries(object).map(([key, value]) =>
+    kv(key, JSON.stringify(value))
+  );
+  return objectify(keyValues);
+}
+
+function combineObjectKeyValuePairs(
+  parts: [key: string, literalValue: string][],
+) {
+  return objectify(parts.map(([key, value]) => kv(key, value)));
+}
+
+function generateObjectCode(inputObject: Subcommand | Arg | Option, init: {
+  path: string[];
+}): string {
+  const objectParts: [key: string, value: string][] = [];
+  for (const [objectKey, objectValue] of Object.entries(inputObject)) {
+    if (objectKey === "action") {
+      continue;
+    }
+    if (objectKey === "subcommands") {
+      const subcommandStrings = (objectValue as Subcommand[])
+        .map((subcommand, index) =>
+          generateObjectCode(subcommand, {
+            path: [...init.path, `s:${index}`],
+          })
+        );
+      objectParts.push([objectKey, arrayify(subcommandStrings)]);
```

Comment:
This common pattern could become
```suggestion
      const subcommands = generateObjectCodeArray(objectValue, (_, index) => ({
        path: [...init.path, `s:${index}`],
      });
      objectParts.push([objectKey, subcommands]);
```

Essentially map over the array to get the init object.